core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1320,6 +1331,175 @@ protected SqlNode performUnconditionalRewrites(
     return node;
   }
 
+  private SqlNode performHiveUdtfRewrite(SqlNode node) {
+    Map<SqlSelect, UdtfInfo> selectUdtfInfo = new HashMap<>();
+     SqlNode rewirte = performHiveUdtfRewriteInternal(node, selectUdtfInfo);
+   //  System.out.println("before:\n" + node.toString() +" \nafter:\n" + rewirte.toString());
+    return node;
+  }
+
+  private SqlNode performHiveUdtfRewriteInternal(SqlNode current,
+                                                 Map<SqlSelect, UdtfInfo> selectUdtfInfos) {
+    System.out.println("start rewrite:" + current);
+
+    // rewrite hive "udtf() as (f0,f1)" to lateral table
+    if (current instanceof SqlSelect) {
+      SqlSelect select = (SqlSelect) current;
+      SqlNodeList newSelectItem =
+          performRewriteForSelectItem(select,
+              select.getSelectList(), selectUdtfInfos);
+
+      UdtfInfo udtfInfo = selectUdtfInfos.get(select);
+      // if the select items contain a table function,
+      // join the from node with the table function.
+      if (udtfInfo != null && select.getFrom() != null) {
+        SqlBasicCall joinRight = createLateralTable(udtfInfo);
+        SqlNode newFrom = new SqlJoin(
```

Comment:
Maybe use `SqlCall` is better.

Reply:
well,the createLateralTable() method return a "as" node, so it must be a SqlBasicCall.